server/src/main/java/org/elasticsearch/index/mapper/FieldMapper.java

```diff
@@ -495,44 +509,43 @@ public MultiFields build(Mapper.Builder mainFieldBuilder, MapperBuilderContext c
                         FieldMapper mapper = entry.getValue().apply(context);
                         mappers.put(key, mapper);
                     }
-                    return new MultiFields(Collections.unmodifiableMap(mappers));
+                    return new MultiFields(mappers);
                 }
             }
         }
 
-        private final Map<String, FieldMapper> mappers;
+        private final FieldMapper[] mappers;
 
         private MultiFields(Map<String, FieldMapper> mappers) {
-            this.mappers = mappers;
+            this.mappers = mappers.values().toArray(EMPTY_MAPPERS);
+            // sort for consistent iteration order + serialization
+            Arrays.sort(this.mappers, Comparator.comparing(FieldMapper::name));
```

Comment:
It looks like there are only 2 usages of this constructor: empty case and a builder one.
Could we change constructor to accept a [sorted] array of field mappers already?
It looks like builder create the maps specifically for this constructor and it could be changed to array/list/stream instead

Reply:
Right ... that makes perfect sense and is yet another neat speedup to constructing these (which is nice for many-shards benchmarks). We only really (sort of) need the map in the builder but not in the constructor.
I made it use an array now for the constructor argument but kept the sorting in the constructor to make things easier to read (otherwise it's yet another mental step to figuring out that the array is sorted).
Let me know what you think :)